Add Slack bot picker

lasryaric opened the pull request:
Using the Slack interactive UI API to allow users to pick up an assistant when talking to our bot if no assistant could be derived from channel based routing rules and the message has no direct mention (eg: ~gpt4).

<img width="844" alt="Screen Shot 2023-10-16 at 11 17 29" src="https://github.com/dust-tt/dust/assets/358965/efbef103-8bd4-46c3-97e1-c4d77b6e11ad">
<img width="946" alt="Screen Shot 2023-10-16 at 11 17 32" src="https://github.com/dust-tt/dust/assets/358965/447757ec-f4c2-46ac-8b4c-bad424458387">
<img width="1005" alt="Screen Shot 2023-10-16 at 11 18 17" src="https://github.com/dust-tt/dust/assets/358965/bd6a45f9-c7aa-4754-b52c-d5d676973a47">

PopDaph commented on connectors/src/api/webhooks/webhook_slack_reaction.ts:
```diff
@@ -0,0 +1,106 @@
+  const agentConfigId =
+    payload.state?.values?.agentConfigId?.static_selectAgentConfig
+      ?.selected_option?.value;
+  console.log("agentConfigId", agentConfigId);
```
NIT: remove debug

PopDaph commented on connectors/src/api_server.ts:
```diff
@@ -26,6 +26,8 @@ import { webhookSlackAPIHandler } from "@connectors/api/webhooks/webhook_slack";
 import logger from "@connectors/logger/logger";
 import { authMiddleware } from "@connectors/middleware/auth";
 
+import { webhookSlackReactionsAPIHandler } from "./api/webhooks/webhook_slack_reaction";
```
NIT: `@connectors/api/webhooks/webhook_slack_reaction`?

PopDaph commented on connectors/src/connectors/slack/bot.ts:
```diff
@@ -42,8 +42,9 @@ export async function botAnswerMessageWithErrorHandling(
-  slackThreadTs: string | null
-): Promise<Result<AgentGenerationSuccessEvent, Error>> {
+  slackThreadTs: string | null,
+  mentionOverride: string[]
```
Why this name? It's never an "override" per say as this can only be filled it there was no mention in the content of the message initially right? 

Also, why an array over string or null? Can we have multiple value in here?

PopDaph commented on connectors/src/api/webhooks/webhook_slack_reaction.ts:
```diff
@@ -0,0 +1,106 @@
+  | null
+  | APIErrorWithStatusCode;
+
+const _webhookSlackReactionsAPIHandler = async (
```
Why is it called Reaction? Is it because it's called like that from Slack API? Could we rename to `Interaction` instead?
Because on the app reaction is for Message Reaction <> Emojis, and I'm afraid this could be confusing.

PopDaph commented on connectors/src/api_server.ts:
```diff
@@ -84,6 +88,10 @@ export function startServer(port: number) {
 
   app.post("/webhooks/:webhook_secret/slack", webhookSlackAPIHandler);
+  app.post(
+    "/webhooks/:webhook_secret/slack_reactions",
```
NIT: `/webhooks/:webhook_secret/slack_reactions` to `/webhooks/:webhook_secret/slack_message_interaction`? Does that make sense?

PopDaph commented on connectors/src/connectors/slack/bot.ts:
```diff
@@ -42,8 +42,9 @@ export async function botAnswerMessageWithErrorHandling(
   slackMessageTs: string,
-  slackThreadTs: string | null
-): Promise<Result<AgentGenerationSuccessEvent, Error>> {
+  slackThreadTs: string | null,
```
NIT: small refactor to have an object over a list of params for this function that gets now 7 params ? 🙏🏻

spolu commented on connectors/src/connectors/slack/bot.ts:
```diff
@@ -589,7 +649,7 @@ async function makeContentFragment(
       return new Err(
         new Error(
-          "Received unexpected undefined replies from Slack API in syncThread (generally transient)"
+          "Received unexpected undefined replies from Slack API in makeContentFragment (generally transient)"
```
generally transient should be removed?

spolu replied:
bump

lasryaric commented:
Per my Slack message, I am working on this PR:

>Just realized that I  hacked around the “edit message” to mention a bot on Friday because I did not have a proper api/message/edit endpoint, so adding that now on top of addressing your comments.

lasryaric commented:
ready for re-review.

spolu commented on connectors/src/connectors/slack/bot.ts:
```diff
@@ -158,6 +267,18 @@ async function botAnswerMessage(
+    apiKey: connector.workspaceAPIKey,
+  });
+
+  const agentConfigurationsRes = await dustAPI.getAgentConfigurations();
```
You're missing one `getAgentConfigurations` to clean-up after moving it here: l.406?

spolu reviewed (changes requested):
Left a small change request. Overall looks good, but the code is starting to be hard to reason about. Would be nice to clean-it up a bit with some additional structure for the different flows of response.

lasryaric commented:
Closing this PR for now as per this Slack thread:
https://dust4ai.slack.com/archives/C050A0S2Z7F/p1698153449669429
